pkg/security/san_pinning.go

```diff
@@ -0,0 +1,183 @@
+	"net/url"
+	"strings"
+
+	"github.com/go-logr/logr"
+
+	"github.com/kyma-project/lifecycle-manager/api/v1alpha1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/runtime-watcher/listener/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	debugLogLevel = 2
+
+	xfccHeader           = "X-Forwarded-Client-Cert"
+	headerValueSeparator = ";"
+	keyValueSeparator    = "="
+	certificateKey       = "Cert="
+
+	shootDomainKey = "skr-domain"
+)
+
+var (
+	// Static errors.
+	errNotVerified   = errors.New("SAN from certificate does not match domain specified in KymaCR")
+	errPemDecode     = errors.New("failed to decode PEM block")
+	errEmptyCert     = errors.New("empty certificate")
+	errHeaderMissing = fmt.Errorf("request does not contain '%s' header", xfccHeader)
+
+	// Error templates.
+	errXFCCParse          = "failed to parse PEM block into x509 certificate: %w"
+	errDecodeURL          = "could not decode certificate URL format: %w"
+	errAnnotationsMissing = "KymaCR '%s' does not have annotation `%s`"
+)
+
+type RequestVerifier struct {
+	Client client.Client
+	log    logr.Logger
+}
+
+func NewRequestVerifier(client client.Client) RequestVerifier {
+	return RequestVerifier{
+		Client: client,
+		log:    ctrl.Log.WithName("listener"),
+	}
+}
+
+// Verify verifies the given request by fetching the KymaCR given in the request payload
+// and comparing the SAN(subject alternative name) of the certificate with the SKR-domain of the KymaCR.
+// If the request can be verified 'nil' will be returned.
+func (v RequestVerifier) Verify(request *http.Request) error {
+	certificate, err := v.getCertificateFromHeader(request)
+	if err != nil {
+		return err
+	}
+
+	domain, err := v.getDomain(request)
+	if err != nil {
+		return err
+	}
+
+	if containsURI(certificate.URIs, domain) ||
+		containsDNS(certificate.DNSNames, domain) ||
+		containsIP(certificate.IPAddresses, domain) {
+		v.log.V(debugLogLevel).Info("Received request verified")
+		return nil
+	}
+	return errNotVerified
+}
+
+// getCertificateFromHeader extracts the XFCC header and pareses it into a valid x509 certificate.
+func (v RequestVerifier) getCertificateFromHeader(r *http.Request) (*x509.Certificate, error) {
+	// Fetch XFCC-Header data
+	xfccValue, ok := r.Header[xfccHeader]
+	if !ok {
+		return nil, errHeaderMissing
+	}
+	xfccData := strings.Split(xfccValue[0], headerValueSeparator)
+
+	// Extract raw certificate
+	var cert string
+	for _, keyValuePair := range xfccData {
+		if strings.Contains(keyValuePair, certificateKey) {
+			cert = strings.Split(keyValuePair, keyValueSeparator)[1]
+			break
+		}
+	}
+	if cert == "" {
+		return nil, errEmptyCert
+	}
+
+	// Decode URL-format
+	decodedValue, err := url.QueryUnescape(cert)
+	if err != nil {
+		return nil, fmt.Errorf(errDecodeURL, err) //nolint:goerr113
+	}
+	decodedValue = strings.Trim(decodedValue, "\"")
+
+	// Decode PEM block and parse certificate
+	block, _ := pem.Decode([]byte(decodedValue))
```

Comment:
The second return value is not an error, ist is the `rest` - meaning if there is  more string than a valid PEM block, it will be output as rest.
But since this could happen, eventho the Decode was successful, we do not need to take care of the rest